File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {

        //Create a hashmap
        Map<Integer, Integer> map = new HashMap<>();

        //Go through each element in arr
        for(int i = 0; i < nums.length;i++){

            int complement = target - nums[i];
            
            if(map.containsKey(complement)){
                return new int[]{map.get(complement),i};
            }

            map.put(nums[i],i);
        }
        return new int[0];
    }
}

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        given an array nums, and a target
        need to return indices of two nums s.t. they add up to target

        a brute force way to do this is using two loops to check each pair and combination
        this would give us an O(n^2) TC

        another way to do this is by subtracting the current number by the target, 
        and check if the diff between the two nums are already seen in the hashmap
        """

        seen = {}

        for i in range(len(nums)):
            diff = target - nums[i]

            if diff in seen:
                return [seen[diff], i]
            
            seen[nums[i]] = i

    